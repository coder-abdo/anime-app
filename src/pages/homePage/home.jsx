import React, { useEffect, useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { fetchTopAnimes, searchAnimes } from "../../redux/actions/fetchAnimes";
import { Animes } from "../../components/animes/Animes";
import { Loader } from "../../components/loader/Loader";
import { AnimesTitle } from "../../sharedStyle/Container";
import { SearchAnimeForm } from "./homePageStyle";
export const Home = () => {
  const { animes } = useSelector(state => state.animes);
  const dispatch = useDispatch();
  const [query, setQuery] = useState("");
  const [title, setTitle] = useState("top 50 animes");
  const handleQueryChange = e => {
    setQuery(e.target.value);
  };
  const handleQuerySubmit = e => {
    e.preventDefault();
    if (query) {
      dispatch(searchAnimes(query));
      setTitle(`${query} results`);
    }
    setQuery("");
  };
  useEffect(() => {
    dispatch(fetchTopAnimes());
    // eslint-disable-next-line
  }, []);
  return (
    <>
      <SearchAnimeForm onSubmit={handleQuerySubmit}>
        <input
          type="text"
          value={query}
          onChange={handleQueryChange}
          placeholder="Search Your Fav Anime"
          data-testid="search-input"
        />
        <button type="submit" data-testid="search-submit-btn">
          search
        </button>
      </SearchAnimeForm>
      <AnimesTitle data-testid="main-title">{title} </AnimesTitle>
      <>{!!animes.length ? <Animes animes={animes} /> : <Loader />}</>
    </>
  );
};
