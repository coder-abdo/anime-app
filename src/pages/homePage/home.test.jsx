import React from "react";
import { Provider } from "react-redux";
import {
  render,
  cleanup,
  waitForElement,
  fireEvent
} from "@testing-library/react";
import { Router } from "react-router";
import { createBrowserHistory } from "history";
import "@testing-library/jest-dom";
import { Home } from "./home";
import { Store } from "../../redux/store/store";
afterEach(cleanup);
describe("render home page with list of animes and without", () => {
  const history = createBrowserHistory();
  test("render home page properly", () => {
    const { container, asFragment } = render(
      <Provider store={Store}>
        <Router history={history}>
          <Home />
        </Router>
      </Provider>
    );
    expect(container).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render loader before list anime", () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router history={history}>
          <Home />
        </Router>
      </Provider>
    );
    expect(getByTestId("loader")).toBeInTheDocument();
  });

  test("should render list of animes if there are list of animes", async () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router history={history}>
          <Home />
        </Router>
      </Provider>
    );
    const animesList = await waitForElement(() => getByTestId("animes"));
    expect(animesList).toBeInTheDocument();
  });
  test("fire submit button properly", () => {
    const { getByTestId } = render(
      <Provider store={Store}>
        <Router history={history}>
          <Home />
        </Router>
      </Provider>
    );
    expect(getByTestId("main-title").textContent).toMatch(/top 50 animes/gi);
    expect(getByTestId("search-input").value).toBe("");
    fireEvent.change(getByTestId("search-input"), {
      target: { value: "hero" }
    });
    expect(getByTestId("search-input").value).toBe("hero");
    fireEvent.click(getByTestId("search-submit-btn"));
    expect(getByTestId("main-title").textContent).toMatch(/hero results/gi);
  });
});
