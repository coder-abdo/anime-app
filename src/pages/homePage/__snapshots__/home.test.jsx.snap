// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render home page with list of animes and without render home page properly 1`] = `
<DocumentFragment>
  <form
    class="sc-AxheI kYvwBi"
  >
    <input
      data-testid="search-input"
      placeholder="Search Your Fav Anime"
      type="text"
      value=""
    />
    <button
      data-testid="search-submit-btn"
      type="submit"
    >
      search
    </button>
  </form>
  <h1
    class="sc-AxiKw eBJYAL"
    data-testid="main-title"
  >
    top 50 animes 
  </h1>
  <div
    class="sc-AxhUy lmioNN"
    data-testid="loader"
  >
    <div
      class="sc-AxgMl cIiRZo"
    />
  </div>
</DocumentFragment>
`;
